### ISSUE TITLE
Push Adguard to Amazon App Store

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#17]
Important things to take into account:
1. Application update - amazon build shouldn't download&apply our apk when update is available
2. We should implement subscriptions within amazon app store.

**Comment by TPS** [comment]
If you all could push the Beta separately (failing a Play-style testing opt-in) & then keep each religiously updated, that'd be appreciated. ☺

**Comment by ameshkov** [comment]
Do you mean push it simply as separate app?
I am not sure if it's allowed by amazon.

**Comment by ameshkov** [comment]
@Revertron has told me that Amazon has pretty similar beta versions feature as Play has. So I guess we can push it there.

**Comment by TPS** [comment]
It _does_?  This is news to me, as I've __never__ known it to be used!  Amazing!  So, then, just 1 app listing, with a beta optin for whoever wants it should suffice.

**Comment by TPS** [comment]
From https://github.com/AdguardTeam/AdguardForAndroid/issues/22#issuecomment-137577596
>  We cannot push the app to amazon until we integrate with their in-app subscriptions.

Ick, then a separate build than the official website 1?  Or just 1 unified 1?  I'm starting to feel sorry for thinking this would be a good thing! :frowning:

**Comment by ameshkov** [comment]
Yep, it will be a separate build, it cannot be unified. We have already done it once for Google Play.

Anyway, we've almost finished with this task:)

**Comment by ameshkov** [comment]
@Revertron, error still happens in non-amazon build:

    Error connecting to subscription service: java.lang.NullPointerException

**Comment by ameshkov** [comment]
I've added null check there but not sure if it's right

**Comment by ameshkov** [comment]
@TPS we've pushed Adguard's beta to amazon app store. If you want to get beta releases we'll need your email.

**Comment by TPS** [comment]
@ameshkov I've sent a note to @Revertron's e-mail for an unrelated issue, so he, being braver than you or I to post _his_ contact info on GitHub, has my address. ;)  & thanks!

**Comment by Revertron** [comment]
@TPS I've taken your address from support ticket ;) Sent an invitation.

**Comment by TPS** [comment]
Oh, yeah, that works, too.… ☺

For the sake of the beta testing I've been doing on what will end up as 3 to 4 tickets _simultaneously_, which beta build is preferred? Direct download or Amazon (as I expect I can't run these in parallel)?

**Comment by Revertron** [comment]
@TPS in fact all we need from amazon testing is the subscription and HD icon for the launcher.
You can test subscription by resetting license on you Fire and trying to purchase it from Amazon. Because you are a Tester Amazon will not charge money for such trial.

If the subscription and HD icon are okay - we are going to publish the "normal" apk to Amazon and consider this task done :)

**Comment by TPS** [comment]
@Revertron I don't own a Fire, but use Amazon Appstore for my general Android device apps, along with GPlay & F-droid, but my only paid apps are via Amazon. Icon looks just fine to _me_, though.

I went as far as hitting subscription option via Amazon, but aborted out before it could charge the $1.50 monthly.  It sure looked like it wanted to charge, so I don't understand
> Because you are a Tester Amazon will not charge money for such trial

Also, I thought there was a $9.95 yearly subscription, but I couldn't find that option.…

**Comment by Revertron** [comment]
@TPS the Fire has it's own launcher with app icons like 512x512 on a main screen. Clearly you can not test it :)

About charging $1.5 - they state that it will look like normal subscription, but without real charging.

About annual subscription - we didn't enable it for now, may be we will enable it in future.

**Comment by TPS** [comment]
@Revertron Yep, worked as mentioned. ☺ Installed Amazon version on my ARM phone: VPN with premium works fine, & not charged for test subscription. Looks good!

Now we need someone with a Fire device (preferably a tablet, as Amazon discontinued the phone last week).

**Comment by ameshkov** [comment]
Let's test subscription purchase one more time and close this issue.

**Comment by TPS** [comment]
My subscription that I "purchased" is still active into next month, so _I_ can't 'til then, I think.

**Comment by TPS** [comment]
Another thing I just noticed: In the app, on the "Manage License" page, there's a note about sharing this single license across all Amazon-installed versions?  If true, can it somehow also be shared across *non*-Amazon-installed versions (perhaps by putting licensing info in online AG account or some such)?  _*That*_ would make that $1.50 a month an awesome deal!  If this isn't what's meant, then maybe reword that?

**Comment by ameshkov** [comment]
As I understand you need an amazon underground app for this to work properly. And if you already have an amazon underground app then why not installing amazon version of Adguard?:)

**Comment by TPS** [comment]
AFAIK, Amazon Underground apps have premium features enabled for users at no additional charge, &, @ this point in time, no charge _@ all_! (I've no understanding of [Amazon's explanation of how anyone gets paid](https://developer.amazon.com/public/solutions/underground) — any1 smarter than me is welcome to try if they promise to use words of ≤ 3 syllables. ;)

I'm not talking about this @ all re: AG!  I just wanted a bit of clarification of the "Manage License" page on the Amazon build. ☺

**Comment by TPS** [comment]
Now AG is prompting to upgrade to build .168 & redirects me to Amazon Appstore, but (I guess ;) without invite link, it doesn't find any update.…  I'll wait! ☺

**Comment by ameshkov** [comment]
Finally, it is available:
http://www.amazon.com/Performix-Adguard-for-Android/dp/B0158CTKR8

**Comment by TPS** [comment]
@ameshkov @Revertron A few notes:

1. I had to purchase from the above link (again), but this is a _separate_ app link than the test at http://www.amazon.com/dp/B0158CTQ2C/ apparently,  though 1 updates the other.

2. The internal updater still wants to update to 2.1.177, which doesn't (yet?) exist on Amazon.

3.  Updating to .176 (which _does_ exist) immediately invalidated my testing license, which is fine, but…

4.  There's no mention of the yearly subscription ([yet?](https://github.com/AdguardTeam/AdguardForAndroid/issues/17#issuecomment-140291937)), which is what I was hoping to enroll.

Should I file separate bugs for 2 & 4?  Or should just exist as a kind of tracking bug?

**Comment by Revertron** [comment]
@TPS The problem is that between builds of amazon and our website version the filters were updated and our build system incremented version.
I submitted the build with new (177) version to amazon, it will resolve the issue.

There is no yearly subscription now.

**Comment by ameshkov** [comment]
> There's no mention of the yearly subscription (yet?), which is what I was hoping to enroll.

Let's file a bug for it, but I can't yet give you an ETA on this.
Do you really think it's important to users to have an annual subscription?

**Comment by TPS** [comment]
@Revertron Thanks for the update! ☺

@ameshkov I've responded in my filed bug, #93 ☺

**Comment by TPS** [comment]
@ameshkov I've received beta update notification directly for https://github.com/AdguardTeam/AdguardForAndroid/releases/tag/v2.5.70-beta for my tablet, but still no word from Amazon for that beta for my phone....  Is it in process from your end?

**Comment by ameshkov** [comment]
Yeah, amazon is slow as always:)

**Comment by TPS** [comment]
Here we go again: 14 hours & counting since https://github.com/AdguardTeam/AdguardForAndroid/tree/v2.6.67-beta, but no update with Amazon.…  I'll update when/if I finally receive it.

**Comment by ameshkov** [comment]
Huh, i hope it will be available today

**Comment by TPS** [comment]
@ameshkov I was reading through https://github.com/AdguardTeam/AdguardForAndroid/issues/17#issuecomment-140069504 because of #568, & it occurred to me that maybe my standing invite that you sent a long time ago might be expired (since still _no_ update! 😢), so can you resend that invite via Amazon?

**Comment by ameshkov** [comment]
@TPS seems that this is our mistake, the new version just was not submitted to beta test (it is in some intermediate "ready to submit" state now)

**Comment by TPS** [comment]
I'll let you know when it pops ↑.

**Comment by TPS** [comment]
Installed & running now.…  Thanks!

**Comment by TPS** [comment]
@ameshkov @Revertron @AdGuardTeam The [ß version](http://smile.amazon.com/dp/B0158CTQ2C/)  is officially _missing_ from Amazon.…

What happened?!

**Comment by ameshkov** [comment]
@TPS it is stuck in "processing" state for some unknown reason. Can't remove or re-upload it. We've sent a message to Amazon support today, but they are not really fast, you know.

**Comment by ameshkov** [comment]
Ah, also while trying to do something, we've removed the previous beta build. It didn't help, but it might've resulted in the total beta absense..

**Comment by TPS** [comment]
Thanks for letting me know.  If I need to do anything different, just tell me.

**Comment by TPS** [comment]
Now Appstore wants to update, but it shows it's going to update _from_ v2.9.70 _to_ v2.9.70, which makes me think it wants to update ß → release, which has generally been *un*recoverable (can't return _to_ ß _from_ release w/o uninstall/reinstall, @ least _last_ time, which is no good w/o ~#119→~ #133).

Can you confirm the version it wants to update _to_ is release v2.9.70 or ß v2.9.111?

**Comment by TPS** [comment]
Thanks, @AdguardTeam, I just got my new Tester invite.

It's updating to ß v2.9.11**0**, though 2.9.11**1** is [detailed as "A hotfix for the v2.9.110 that contained a critical error."](https://github.com/AdguardTeam/AdguardForAndroid/releases/tag/v2.9.111-beta)  Based on _that_, I don't want to actually complete the update, just to cripple my AG installation!

Please advise!

**Comment by TPS** [comment]
Thanks, @AdguardTeam, I've now v2.9.111, & all is well. :bow:
